plugins/catalog-backend/src/ingestion/processors/UrlReaderProcessor.ts

```diff
@@ -22,42 +22,77 @@ import { Logger } from 'winston';
 import * as result from './results';
 import {
   CatalogProcessor,
+  CatalogProcessorCache,
   CatalogProcessorEmit,
+  CatalogProcessorEntityResult,
   CatalogProcessorParser,
+  CatalogProcessorResult,
 } from './types';
 
+const CACHE_KEY = 'v1';
+
 type Options = {
   reader: UrlReader;
   logger: Logger;
 };
 
+type CacheItem = {
+  etag: string;
+  value: CatalogProcessorEntityResult[];
+};
+
 export class UrlReaderProcessor implements CatalogProcessor {
   constructor(private readonly options: Options) {}
 
+  getProcessorName() {
+    return 'url-reader';
+  }
+
   async readLocation(
     location: LocationSpec,
     optional: boolean,
     emit: CatalogProcessorEmit,
     parser: CatalogProcessorParser,
+    cache: CatalogProcessorCache,
   ): Promise<boolean> {
     if (location.type !== 'url') {
       return false;
     }
 
+    const cacheItem = await cache.get<CacheItem>(CACHE_KEY);
+
     try {
-      const output = await this.doRead(location.target);
+      const [output, newEtag] = await this.doRead(
+        location.target,
+        cacheItem?.etag,
+      );
+
+      const parseResults: CatalogProcessorResult[] = [];
       for (const item of output) {
         for await (const parseResult of parser({
           data: item.data,
           location: { type: location.type, target: item.url },
         })) {
+          parseResults.push(parseResult);
           emit(parseResult);
         }
       }
+
+      const isOnlyEntities = parseResults.every(r => r.type === 'entity');
+      if (newEtag && isOnlyEntities) {
+        await cache.set<CacheItem>(CACHE_KEY, {
+          etag: newEtag,
+          value: parseResults as CatalogProcessorEntityResult[],
```

Comment:
Yeah this is a bit scary, as we evolve a type like `CatalogProcessorEntityResult` we also simultaneously have to remember to bump the cache version keys anywhere that uses that type, either directly or indirectly. And that will apply inside third party processors that are out of our control as well. I wonder if we should bump the prio of getting type generation + validation into place soon after this, and just silently reject read entries that don't validate against the expected schema

Reply:
Yeah I think the processor won't often even _know_ that there's been a structural change to one of the types involved in the cache data. This is something that the legacy java backstage-backend has struggled _a lot_ with so the problem is very real.

Almost like we'd want to have a `get<T>(key: string, type: KeyType<T>): Promise<T>` kinda like [jackson type references](https://fasterxml.github.io/jackson-core/javadoc/2.8/com/fasterxml/jackson/core/type/TypeReference.html)? :) That would be a place for us to place jsonschema validation and serde.

Not sure that unconditional wipe on error is always better either? Hmm. Tricky. If the error is a rate limiting error then wiping out an etag or whatnot will probably make the problem worse.

Reply:
Gone with inlining the type along with a warning for now, and we can look at more in-depth validation in the future